File: solutions/combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        def findCombination(i, cur, total):
            if total == target:
                res.append(cur.copy())
                return
            if i >= len(candidates) or total > target:
                return
            cur.append(candidates[i])
            findCombination(i, cur, total + candidates[i])
            cur.pop()
            findCombination(i+1, cur, total)

        findCombination(0, [], 0)
        return res

File: solutions/combination-sum/combination-sum.go
func combinationSum(candidates []int, target int) [][]int {
    var res [][]int
    dfs(0, []int{}, 0, candidates, &res, target)
    return res
}

func dfs(i int, cur []int, total int,candidates []int, res *[][]int, target int) {
    if total == target {
        *res = append(*res, append([]int{}, cur...))
        return
    }
    if (i >= len(candidates)) || (total > target) {
        return
    }

    cur = append(cur, candidates[i])
    dfs(i, cur, total + candidates[i], candidates, res, target)
    cur = cur[:len(cur)-1]
    dfs(i+1, cur, total, candidates, res, target)
}